File: leetcode/139.py
class Solution(object):
    def wordBreak(self, s, dict):
        n, m, G = len(s), len(dict), {}
        f = [False] * (n + 1)
        f[0] = True
        for ss in dict: G[ss] = 1
        for i in xrange(1, n + 1):
            for j in xrange(0, i):
                if f[j] and s[j : i] in G:  
                    f[i] = True
                    break
        return f[n]

File: leetcode/139.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& S) {
        unordered_map<string, bool> G;
        for (auto& t : S)   G[t] = 1;
        for (int i = 0; i < s.size(); ++i)
            for (int j = 0; j < i; ++j) {
                if (G[s.substr(0, j + 1)] && G[s.substr(j + 1, i - j)])  G[s.substr(0, i + 1)] = 1;
            }
        return G[s];
    }
};